### DISCUSSION TITLE
What might be a good way to test entering an infinite loop?

### DISCUSSION THREAD

**Discussion by EricLauber** [bxparks/AUnit#120] [Q&A]
I am building a "configuration mode" into my Arduino program. When the device first powers on, it'll use a button-press to determine if it should enter "configuration mode" or continue to normal operation. Configuration mode is a state machine that uses short presses and long presses to advance through the machine. The is no way to "exit" the state machine apart from an option that calls `ESP.restart()`. It's intentionally an infinite loop.

Startup for my program amounts to:

```
if (buttonPressDetected)
{
    callConfigModeInfiniteLoop();
}
else
{
    doTheOtherStuff();
}
```

Since there's a conditional statement and some other logic that happens, I want to write tests to prove this works. Using AUnit (and EpoxyDuino although I'm not sure if that matters here) I'm able to successfully go down both paths. Except - it's an infinite loop, so for the configuration mode path, tests lock up.

Is there a way I can force the test to end after a period of time? Maybe the loop can write to serial output and AUnit can end the test and read the output? Or in the loop function but prior to it actually starting I could set some variable in my class that AUnit could detect...

In other languages I might have built an event that the loop fires, and the test would subscribe to the event. By subscribing to the event, I could interrupt the loop, assert some value of the class (or the fact that the event fired at all is a positive assertion), and end the test.

**Comment by bxparks** [comment]
Maybe I'm misunderstanding something, but why do you need to test such a simple if/then/else code? You don't have to test every little bit of your code...

With regards to timeout, I ~~think~~ don't think there is anything that AUnit can help you with here. AUnit is like any other testing/assertion framework. It assumes that the thing you are testing is a (relatively) self-contained piece, which can be probed externally with various inputs, and AUnit helps you assert that the output is what you expect. If you have something inside `callConfigModeInfiniteLoop()` that you want to test, you need to split those pieces into testable units, and test those separately. Testing an infinite loop is.. odd. I've never had to do it.
